firedrake/interpolation.py

```diff
@@ -227,17 +242,55 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
         raise RuntimeError('Shape mismatch: Expression shape %r, FunctionSpace shape %r'
                            % (expr.ufl_shape, V.ufl_element().value_shape()))
 
-    mesh = V.ufl_domain()
-    coords = mesh.coordinates
+    target_mesh = V.ufl_domain()
+    try:
+        trans_mesh = (
+            expr.ufl_domain() is not None  # Constant expr
+            and expr.ufl_domain() != V.mesh()  # Coming from a different domain
+        )
+    except AttributeError:
+        # Have python Expression
+        trans_mesh = False
+
+    if trans_mesh:
+        if target_mesh.geometric_dimension() != expr.ufl_domain().geometric_dimension():
+            raise ValueError("Cannot interpolate onto a mesh of a different geometric dimension")
+        # In the trans-mesh case we loop over the target mesh cells so we need
+        # to compose a map that takes us from their cells to the source mesh
+        # cells, and (where necessary) to the function space nodes on the
+        # source mesh.
+        if isinstance(target_mesh.topology, firedrake.mesh.VertexOnlyMeshTopology):
+            target_mesh_source_mesh_map = target_mesh.cell_parent_cell_map
+        else:
+            raise NotImplementedError("Can only interpolate onto a Vertex Only Mesh")
+        # For trans-mesh interpolation we use a FInAT QuadratureElement as the
+        # (base) target element with runtime point set expressions as their
+        # quadrature rule point set and weights from their dual basis.
+        # NOTE: This setup is useful for thinking about future design - in the
+        # future this `rebuild` function can be absorbed into FInAT as a
+        # transformer that eats an element and gives you an equivalent (which
+        # may or may not be a QuadratureElement) that lets you do run time
+        # tabulation. Alternatively (and this all depends on future design
+        # decision about FInAT how dual evaluation should work) the
+        # to_element's dual basis (which look rather like quadrature rules) can
+        # have their pointset(s) directly replaced with run-time tabulated
+        # equivalent(s) (i.e. finat.point_set.UnknownPointSet(s))
+        rt_var_name = 'rt_X'
+        to_element = rebuild(to_element, expr, rt_var_name)
+        # The source mesh is on the expression not the FunctionSpace/Function
+        # we are interpolating into/onto
+        source_mesh = expr.ufl_domain()
+    else:
+        # The source mesh is that of the FunctionSpace/Function we are
+        # interpolating into/onto
+        source_mesh = target_mesh
```

Comment:
(Suggested) The logic used in the above (ll. 245 - 286) is slightly confusing to me. I would do:
```
target_mesh = V.ufl_domain()
source_mesh = expr.ufl_domain() or target_mesh
if source_mesh is not target_mesh:
    assert isinstance(target_mesh.topology, firedrake.mesh.VertexOnlyMeshTopology)
    assert target_mesh._parent_mesh is source_mesh
    rt_var_name = 'rt_X'
    to_element = rebuild(to_element, expr, rt_var_name)
```
You need to handle `except AttributeError` part if you need to. Also, I would not introduce new variable for `target_mesh.cell_parent_cell_map`. Maybe we don't need `trans_mesh`, either.

Reply:
I agree